### PR TITLE
ShovelUtilsTW

### PR DESCRIPTION
**Pull request by TheShovel**
TurboWarp version of my misc utilities extension.

It includes:

-Blocks for importing/loading stuff from links
image

-Blocks for setting lists to a json array, and getting their data
image

-And other misc blocks
image

### CODE REVIEW

**Comment by TheShovel** [comment]
Well that's weird. Why is it saying that stuff is not defined when I did define them?

**Comment by softedco** [comment]
> Well that's weird. Why is it saying that stuff is not defined when I did define them?

You can't define stuff as null because null is basically the same thing as undefined.

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 1]
```diff
@@ -0,0 +1,257 @@
+console.log("ShovelUtils v1.1")
+tempImageLoad = null
```
Comment (softedco) [inline extensions/TheShovel/ShovelUtilsTW.js:2]:
```suggestion
var tempImageLoad = {};
```
fixes tempImageLoad is not defined
also it appears to be completely unused

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+string = null;
```
Comment (softedco) [inline extensions/TheShovel/ShovelUtilsTW.js:3]:
```suggestion
var string = '';
```
fixes string is not defined
also unused

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+temp = null;
```
Comment (softedco) [inline extensions/TheShovel/ShovelUtilsTW.js:4]:
```suggestion
var temp = {};
```
fixes temp is not defined

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+tempVAR = null;
```
Comment (softedco) [inline extensions/TheShovel/ShovelUtilsTW.js:5]:
```suggestion
var tempVAR = {};
```
fixes tempVAR is not defined
also unused

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+
+
+(function(Scratch) {
```
Comment (softedco) [inline extensions/TheShovel/ShovelUtilsTW.js:9]:
```suggestion
(function(Scratch) {
  'use strict';
```
fixes use the function form of use strict

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+  const vm = Scratch.vm;
+  'use strict';
```
Comment (softedco) [inline extensions/TheShovel/ShovelUtilsTW.js:11]:
```suggestion

```
I guess it shouldn't be there

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+ */
+brightnessByColor ({color}) {
+  var color = "" + color, isHEX = color.indexOf("#") == 0, isRGB = color.indexOf("rgb") == 0;
+  if (isHEX) {
+    const hasFullSpec = color.length == 7;
+    var m = color.substr(1).match(hasFullSpec ? /(\S{2})/g : /(\S{1})/g);
+    if (m) var r = parseInt(m[0] + (hasFullSpec ? '' : m[0]), 16), g = parseInt(m[1] + (hasFullSpec ? '' : m[1]), 16), b = parseInt(m[2] + (hasFullSpec ? '' : m[2]), 16);
+  }
+  if (isRGB) {
+    var m = color.match(/(\d+){3}/g);
```
Comment (softedco) [inline extensions/TheShovel/ShovelUtilsTW.js:247]:
```suggestion
    m = color.match(/(\d+){3}/g);
```
fixes m is already defined

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    if (m) var r = m[0], g = m[1], b = m[2];
```
Comment (softedco) [inline extensions/TheShovel/ShovelUtilsTW.js:248]:
```suggestion
    if (m) r = m[0], g = m[1], b = m[2];
```
fixes r is already defined

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+  vm.setEditingTarget(vm.runtime.getSpriteTargetByName(NAME).id); 
+}
+
+/**
+ * Calculate brightness value by RGB or HEX color.
+ * @param color (String) The color value in RGB or HEX (for example: #000000 || #000 || rgb(0,0,0) || rgba(0,0,0,0))
+ * @returns (Number) The brightness value (dark) 0 ... 255 (light)
+ */
+brightnessByColor ({color}) {
+  var color = "" + color, isHEX = color.indexOf("#") == 0, isRGB = color.indexOf("rgb") == 0;
```
Comment (softedco) [inline extensions/TheShovel/ShovelUtilsTW.js:240]:
```suggestion
  color = "" + color;
  var isHEX = color.indexOf("#") == 0, isRGB = color.indexOf("rgb") == 0;
```
fixes color is already defined

**Review by softedco — CHANGES_REQUESTED** [review]
this should probably do

**Comment by Mickmick2012** [comment]
looks good ! I was hopping to see this

**Comment by TheShovel** [comment]
Thanks a lot for fixing the issue @softedco :D

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 10]
```diff
@@ -0,0 +1,258 @@
+console.log("ShovelUtils v1.1")
+var tempImageLoad = {};
```
Comment (GarboMuffin) [inline extensions/TheShovel/ShovelUtilsTW.js:2]: All of these variables need to go underneath `(function(Scratch) {` below so they don't pollute the global scope

**Comment by TheShovel** [comment]
Oh god I just realized what those buttons do. I am sorry lmao.

File: extensions/TheShovel/ShovelUtilsTW.js [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+    console.log("Error", error);
+  });
+}
+
+restartProject(){
+vm.greenFlag()
+}
+
+loadExtension({TEXT}){
+  vm.extensionManager.loadExtensionURL(TEXT)
```
Comment (GarboMuffin) [inline extensions/TheShovel/ShovelUtilsTW.js:216]: wrap this in `if (await vm.securityManager.canLoadExtensionFromProject(url)) {`

**Comment by TheShovel** [comment]
shovelutils

Image

**Pull request merged**